table/block_based/mock_block_based_table.h

```diff
@@ -29,11 +31,15 @@ class MockBlockBasedTableTester {
   InternalKeyComparator icomp_;
   std::unique_ptr<BlockBasedTable> table_;
 
-  MockBlockBasedTableTester(const FilterPolicy *filter_policy)
+  MockBlockBasedTableTester(const FilterPolicy* filter_policy)
+      : MockBlockBasedTableTester(
+            std::shared_ptr<const FilterPolicy>(filter_policy)){};
+
+  MockBlockBasedTableTester(std::shared_ptr<const FilterPolicy> filter_policy)
       : ioptions_(options_),
         env_options_(options_),
         icomp_(options_.comparator) {
-    table_options_.filter_policy.reset(filter_policy);
+    table_options_.filter_policy = std::move(filter_policy);
```

Comment:
Any subtle reason why we are doing this trick?

Reply:
`table_options_.filter_policy.reset(filter_policy.get());` would be wrong for shared_ptr, because it would be managed twice (double free). `table_options_.filter_policy = filter_policy;` would be OK but might unnecessarily increment+decrement the refcount on the shared_ptr (due to two potentially live local copies). `std::move` essentially guarantees the optimization by ensuring only one live local copy.

I know @mrambacher likes to use `const std::shared_ptr&` parameters, but in cases like this where the callee will surely (or almost surely) save a copy, I prefer `std::shared_ptr` because you can pass in a temporary like `std::make_shared<X>` and likely avoid unnecessary extra increment+decrement the refcount.

Reply:
Discussed offline: 

For clarification: my original question was intended to ask why we do an "indirection" from raw pointer "const FilterPolicy* filter_policy" to the share pointer "table_options_.filter_policy" using your trick of adding another constructor (I realized I quoted the wrong line for my question so probably that's why it missed its point)

Update:
It's not needed in this code but might be needed for future refactoring. If that's the case, I suggest adding the "indirection" in that PR for better context.

Reply:
OK, I'll take it out of this one. I just noticed I need to fix some FB lints also.